src/rgb/contract/assignments.rs

```diff
@@ -708,3 +708,813 @@ mod strict_encoding {
+    }
+
+    #[test]
+    fn test_zero_balance() {
+        let mut rng = thread_rng();
+
+        // test equal inputs and outputs
+        // 4 values equally distributed among ours and theirs allocations
+        // * One zero value case
+        // * All zero value case
+        // * One u64::MAX value case
+        // * All u64::MAX value case
+        // * One zero value + One u64::MAX value case
+        let amounts = vec![
+            [1u64, 2u64, 3u64, u64::MAX],
+            [10u64, 20u64, 30u64, 40u64],
+            [0u64, 197642u64, u64::MAX, 476543u64],
+            [0u64, 0u64, 0u64, 0u64],
+            [u64::MAX, u64::MAX, u64::MAX, u64::MAX],
+        ];
+
+        for vec in amounts.iter() {
+            assert!(compute_zero_balance(vec, vec, 2));
+        }
+
+        // Test when ours is empty
+        assert!(compute_zero_balance(&amounts[1], &amounts[1], 0));
+
+        // Test when theirs is empty
+        assert!(compute_zero_balance(
+            &amounts[1],
+            &amounts[1],
+            amounts[1].len()
+        ));
+
+        // Test random inputs and outputs
+        // Randomly distributed between ours and theirs allocation
+        for _ in 0..10 {
+            // Randomly generate number of amounts between 1 to 20
+            let length = rng.gen_range(1, 20);
+
+            // Randomly fill the amount vector
+            let mut amounts = vec![0; length];
+            for _ in 0..length {
+                // keep the amount value low for faster testing
+                amounts.push(rng.gen_range::<u64>(0, 10));
+            }
+
+            // Create an output amount vector such that
+            // input.sum() = output.sum(), but
+            // input.count() != output.count()
+            // We do this by generating an empty output vector
+            // of random size. Then randomly add 1 to random positions
+            // until the sum of the vector matches with the input sum
+            // This also creates 0 value allocations probabilistically
+
+            let input_sum: u64 = amounts.iter().sum();
+            let mut output_amounts = vec![0u64; rng.gen_range(1, 20)];
+            let output_length = output_amounts.len();
+            for _ in 0..input_sum {
+                output_amounts[rng.gen_range(0, output_length)] += 1u64;
+            }
+
+            // Check if test passes
+            assert!(compute_zero_balance(
+                &amounts[..],
+                &output_amounts[..],
+                rng.gen_range(0, output_length)
+            ));
+        }
+
+        // Test when input.sum() != output.sum()
+        // When they only differ by 1
+        // When they differ by u64::MAX
+        assert!(!compute_zero_balance(&amounts[1], &amounts[2], 2));
+        assert!(!compute_zero_balance(&[1, 2, 3, 4], &[1, 2, 3, 5], 2));
+        assert!(!compute_zero_balance(
+            &[1, 2, 3, 0],
+            &[1, 2, 3, u64::MAX],
```

Comment:
This test must fail b/c of different reason: amounts can't be larger than `u64::MAX` in their sum, i.e. this line must be equal to 1+2+3-1=5. I think we need to add special test cases to cover overflow situations.

Reply:
Hmm i didn't thought about this. Dont we really need to make sure somehow that overflow in confidential amounts doesn't happen? Right now we are getting overflows i think, multiple `u64::MAX` are adding up together and commit sum verification passing.

Reply:
We can't prevent it from happening; however what we can make sure is that the total supply will never be more than `u64::MAX` (which is done at the level of schema), so the sum of all inputs or outputs can't overflow in practice for the assets